pkg/cmd/build/translator.go

```diff
@@ -0,0 +1,258 @@
+
+	scanner := bufio.NewScanner(readerFile)
+	datawriter := bufio.NewWriter(writerFile)
+	defer datawriter.Flush()
+
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		result := line
+		for _, translator := range dt.translators {
+			result = translator.translate(result)
+		}
+
+		_, err := datawriter.WriteString(result + "\n")
+		if err != nil {
+			return fmt.Errorf("failed to write dockerfile: %w", err)
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		return err
+	}
+	return nil
+}
+
+type translator interface {
+	translate(line string) string
+}
+
+type registryTranslator struct {
+	replacer        registry.Replacer
+	userNs          string
+	globalNamespace string
+}
+
+func newRegistryTranslator(okCtx OktetoContextInterface) registryTranslator {
+	globalNamespace := constants.DefaultGlobalNamespace
+	ctxGlobalNamespace := okCtx.GetGlobalNamespace()
+	if ctxGlobalNamespace != "" {
+		globalNamespace = ctxGlobalNamespace
+	}
+
+	return registryTranslator{
+		replacer:        registry.NewRegistryReplacer(GetRegistryConfigFromOktetoConfig(okCtx)),
+		userNs:          okCtx.GetNamespace(),
+		globalNamespace: globalNamespace,
+	}
+}
+
+func (rt registryTranslator) translate(line string) string {
+	if strings.Contains(line, constants.DevRegistry) {
+		result := rt.replacer.Replace(line, constants.DevRegistry, rt.userNs)
+		return result
+	}
+
+	if strings.Contains(line, constants.GlobalRegistry) {
+		result := rt.replacer.Replace(line, constants.GlobalRegistry, rt.globalNamespace)
+		return result
+	}
+	return line
+}
+
+type cacheMountTranslator struct {
+	projectHash        string
+	cacheMountRegex    *regexp.Regexp
+	hasIDRegex         *regexp.Regexp
+	targetExtractRegex *regexp.Regexp
+	mountRegex         *regexp.Regexp
+	targetRegex        *regexp.Regexp
+}
+
+func newCacheMountTranslator(repo, dockerfilePath, target string) cacheMountTranslator {
+	cacheMountRegex := regexp.MustCompile(`^RUN.*--mount=.*type=cache`)
+	hasIDRegex := regexp.MustCompile(`^RUN.*--mount=[^ ]*id=`)
+	targetExtractRegex := regexp.MustCompile(`--mount=[^ ]*target=([^, ]+)`)
+	mountRegex := regexp.MustCompile(`--mount=([^[:space:]]+)`)
+	targetRegex := regexp.MustCompile(`target=([^,\s]+)`)
+
+	return cacheMountTranslator{
+		projectHash:        generateProjectHash(repo, dockerfilePath, target),
+		cacheMountRegex:    cacheMountRegex,
+		hasIDRegex:         hasIDRegex,
+		targetExtractRegex: targetExtractRegex,
+		mountRegex:         mountRegex,
+		targetRegex:        targetRegex,
+	}
+}
+
+func generateProjectHash(repositoryURL, manifestName, path string) string {
+	// Create input string for hashing
+	input := fmt.Sprintf("%s-%s-%s", repositoryURL, manifestName, path)
+
+	// Generate SHA256 hash
+	hasher := sha256.New()
+	hasher.Write([]byte(input))
+	hash := hasher.Sum(nil)
+
+	// Return first 12 characters of hex hash for readability
+	return hex.EncodeToString(hash)[:12]
+}
+
+func (cmt cacheMountTranslator) translate(line string) string {
```

Comment:
This is the main change. The other things is moving code to allow having the id here and refactoring

Reply:
Could you provide some values for repo, folder and target?
I guess repo is the repo name, like `app` in https://github.com/okteto/app
Folder is the folder of the dockerfile? I would use the relative path of the dockerfile respect to the repo root folder.
Why do we need `target`? what is it?
Fianlly, what's the value of `repo` if the ddockerfile doesn't belong to a git repo?